src/lambda_function/raw_sync/app.py

```diff
@@ -0,0 +1,518 @@
+           recursively with a larger range size. This scenario is expected to be rare,
+           only occurring if there is a comment at the end of the ZIP file exceeding
+           `range_size` - 22 bytes (the size of the EOCD record minus the optional comment).
+           If a non-zip file was provided, this process will repeat until the entire file
+           has been read, since this is the only way to determine that there is no EOCD record.
+
+        2. Having found the EOCD record, the offset of the central directory is determined.
+           If the central directory is not fully contained within the retrieved range,
+           the function will call itself with the appropriate `range_size`.
+
+        3. The function then reads the central directory to extract the file list.
+
+    Args:
+        s3_client (boto3.client): The Boto3 S3 client used to fetch the object from S3.
+        bucket (str): The name of the S3 bucket where the object is stored.
+        key (str): The key of the S3 object containing the ZIP archive.
+        range_size (int): The number of bytes to fetch from the tail of the S3 object on each call.
+                          Defaults to 64 KB.
+
+    Returns:
+        list[dict]: A list of dict with information about the files contained within the ZIP archive.
+            The dict has keys `filename` and `file_size`, which contain the respective values from
+            the ZipInfo object.
+
+            Files are filtered to exclude:
+
+            - Directories (i.e., paths containing "/").
+            - Files named "Manifest".
+            - Empty files (file size == 0).
+
+            If no files match the criteria or the EOCD record is not found, an empty list is returned.
+
+    Notes:
+        - The function may trigger multiple recursive calls if the EOCD record is large or non-existent,
+            but is guaranteed to terminate upon retrieving the entire object.
+    """
+    file_list = []
+    object_response = s3_client.get_object(
+        Bucket=bucket, Key=key, Range=f"bytes=-{range_size}"
+    )
+    logger.debug(f"Object Response: {object_response}")
+
+    # Parse the ContentRange for later reference
+    range_start, range_end, total_size = parse_content_range(
+        content_range=object_response["ContentRange"]
+    )
+
+    # Determine end of central directory offset
+    tail = object_response["Body"].read()
+    eocd_offset = determine_eocd_offset(
+        body=tail, content_range=object_response["ContentRange"]
+    )
+    if eocd_offset == -1:
+        adjusted_range_size = range_size * 2
+        if adjusted_range_size > total_size * 2:
+            logger.error(
+                "Did not find an end of central directory record in "
+                f"s3://{bucket}/{key}"
+            )
+            return []
+        logger.warning(
+            f"Calling this function recursively with `range_size` = {adjusted_range_size}"
+        )
+        return list_files_in_archive(
```

Comment:
Given the current recursion limit in python being 1000, in the scenario that we hit it, are we concerned about `range_size` being multiplied 1000 times/large?

Reply:
That would probably mean a very large file (which is almost certainly not a zip file) was dropped in the input bucket. How large? 32*2^1000 KB, or 3.5e290 petabytes, much larger than any object in S3 could ever be.

So we will never reach the recursion limit and would instead timeout after 15 minutes of trying to download enough of the object to find the EOCD marker.

Reply:
Could there be a scenario where `eocd_offset` was determined incorrectly (-1) for a non-large file?

Reply:
It can't be a zip file without the EOCD byte marker, so it's not possible to have a false negative. Is that what you mean?

Reply:
Yup